app/code/Magento/GiftMessage/Test/Unit/Observer/SalesEventOrderItemToQuoteItemObserverTest.php

```diff
@@ -0,0 +1,196 @@
+
+    /**
+     * Stub new message id
+     */
+    private const STUB_NEW_MESSAGE_ID = 2;
+
+    /**
+     * @var SalesEventOrderItemToQuoteItemObserver
+     */
+    private $observer;
+
+    /**
+     * @var MessageFactory|MockObject
+     */
+    private $messageFactoryMock;
+
+    /**
+     * @var MessageHelper|MockObject
+     */
+    private $giftMessageHelperMock;
+
+    /**
+     * @var Observer|MockObject
+     */
+    private $observerMock;
+
+    /**
+     * @var Event|MockObject
+     */
+    private $eventMock;
+
+    /**
+     * @var Order|MockObject
+     */
+    private $orderMock;
+
+    /**
+     * @var OrderItem|MockObject
+     */
+    private $orderItemMock;
+
+    /**
+     * @var Store|MockObject
+     */
+    private $storeMock;
+
+    /**
+     * @var MessageInterface|MockObject
+     */
+    private $messageMock;
+
+    /**
+     * @var QuoteItem|MockObject
+     */
+    private $quoteItemMock;
+
+    /**
+     * Prepare environment for test
+     */
+    public function setUp(): void
+    {
+        $this->messageFactoryMock = $this->createMock(MessageFactory::class);
+        $this->giftMessageHelperMock = $this->createMock(MessageHelper::class);
+        $this->observerMock = $this->createMock(Observer::class);
+        $this->eventMock = $this->createPartialMock(Event::class, ['getOrderItem', 'getQuoteItem']);
+        $this->orderItemMock = $this->createPartialMock(
+            OrderItem::class,
+            ['getOrder', 'getStoreId', 'getGiftMessageId']
+        );
+        $this->quoteItemMock = $this->createPartialMock(QuoteItem::class, ['setGiftMessageId']);
+        $this->orderMock = $this->createPartialMock(Order::class, ['getReordered']);
+        $this->storeMock = $this->createMock(Store::class);
+        $this->messageMock = $this->createMock(MessageModel::class);
+
+        $objectManager = new ObjectManager($this);
+
+        $this->observer = $objectManager->getObject(
+            SalesEventOrderItemToQuoteItemObserver::class,
+            [
+                'messageFactory' => $this->messageFactoryMock,
+                'giftMessageMessage' => $this->giftMessageHelperMock
+            ]
+        );
+    }
+
+    /**
+     * Tests duplicating gift message from order item to quote item
+     *
+     * @param bool $orderIsReordered
+     * @param bool $isMessagesAllowed
+     * @dataProvider giftMessageDataProvider
+     */
+    public function testExecute($orderIsReordered, $isMessagesAllowed)
+    {
+        $this->eventMock->expects($this->atLeastOnce())
+            ->method('getOrderItem')
+            ->willReturn($this->orderItemMock);
+
+        $this->orderItemMock->expects($this->atLeastOnce())
+            ->method('getOrder')
+            ->willReturn($this->orderMock);
+
+        $this->observerMock->expects($this->atLeastOnce())
+            ->method('getEvent')
+            ->willReturn($this->eventMock);
+
+        if (!$orderIsReordered && $isMessagesAllowed) {
```

Comment:
Based on these 2 conditions you should create at least 4 test methods. For example:
- `public function testReorderWhenGiftMessageIsAllowed()`
- `public function testNewOrderWhenGiftMessageIsAllowed()`
- `public function testReorderWhenGiftMessageIsNotAllowed()`
- `public function testNewOrderWhenGiftMessageIsNotAllowed()`